### DISCUSSION TITLE
Parameter position from `LocalPointerKey` in the SSA IR.

### DISCUSSION THREAD

**Discussion by khatchad** [wala/WALA#1277] [Q&A] [answered]
How do you obtain (consistently and uniformly) the parameter position of a `LocalPointerKey` representing a parameter? A `LocalPointerKey` consists of a value number and a CG node.

I see this code:

https://github.com/wala/WALA/blob/89de874efbc3369097e03b85493f4a3beb8e1a44/core/src/main/java/com/ibm/wala/ipa/callgraph/propagation/LocalPointerKey.java#L62-L64

From this code, we see that there is some correspondence between value numbers and parameters. If `isParameter()` returns true, what can we say about the value number of the `LocalPointerKey`? Can we say that it is the $x-1$ th parameter?

**Comment by msridhar** [comment, marked as answer]
I believe that for instance methods, value number 1 is the receiver, and then the remaining parameters get consecutive value numbers after that.  For static methods, value number 1 is the first declared formal parameter and they are numbered consecutively after that.  Does that match what you are seeing?

**Comment by khatchad** [comment]
No, but I am looking at IR from Python CAst. There, at least for functions (not methods), value number 1 is actually the function itself. I believe that this is how @juliandolby is dealing with the import aliasing in Python.

**Reply by msridhar** [reply]
Ah, ok, I don't really know the scheme for Python.  For JavaScript something similar happens, where value number 1 is the function itself.  So I guess the real answer here is that the scheme may depend on the front-end language being used.

**Comment by khatchad** [comment]
Is this scheme documented anywhere? I had a quick look in the wiki but couldn't find anything.
